ui/src/components/VideoOverlay.tsx

```diff
@@ -398,124 +399,126 @@ export function PointerLockBar({ show }: PointerLockBarProps) {
 interface RebootingOverlayProps {
   readonly show: boolean;
   readonly postRebootAction: PostRebootAction;
+  readonly deviceId?: string; // Required for cloud mode to build versioned URLs
 }
 
-export function RebootingOverlay({ show, postRebootAction }: RebootingOverlayProps) {
+export function RebootingOverlay({ show, postRebootAction, deviceId }: RebootingOverlayProps) {
   const { peerConnectionState } = useRTCStore();
-
-  // Check if we've already seen the connection drop (confirms reboot actually started)
   const [hasSeenDisconnect, setHasSeenDisconnect] = useState(
-    ['disconnected', 'closed', 'failed'].includes(peerConnectionState ?? '')
+    ["disconnected", "closed", "failed"].includes(peerConnectionState ?? ""),
   );
-
-  // Track if we've timed out
   const [hasTimedOut, setHasTimedOut] = useState(false);
+  const isCheckingRef = useRef(false);
+  const abortControllerRef = useRef<AbortController | null>(null);
 
-  // Monitor for disconnect after reboot is initiated
+  // Detect connection drop (confirms reboot started)
   useEffect(() => {
-    if (!show) return;
-    if (hasSeenDisconnect) return;
-
-    if (['disconnected', 'closed', 'failed'].includes(peerConnectionState ?? '')) {
-      console.log('hasSeenDisconnect', hasSeenDisconnect);
+    if (!show || hasSeenDisconnect) return;
+    if (["disconnected", "closed", "failed"].includes(peerConnectionState ?? "")) {
       setHasSeenDisconnect(true);
     }
   }, [show, peerConnectionState, hasSeenDisconnect]);
 
-  // Set timeout after 30 seconds
+  // Timeout after 30 seconds
   useEffect(() => {
     if (!show) {
       setHasTimedOut(false);
       return;
     }
-
-    const timeoutId = setTimeout(() => {
-      setHasTimedOut(true);
-    }, 30 * 1000);
-
-    return () => {
-      clearTimeout(timeoutId);
-    };
+    const id = setTimeout(() => setHasTimedOut(true), 30_000);
+    return () => clearTimeout(id);
   }, [show]);
 
+  // Redirect helper - navigates and forces reload
+  const redirectTo = useCallback(async (url: string) => {
+    window.location.href = url;
```

Comment:
Glad you extracted this pattern out to a utils.ts function... I think I've see it before ;)